### PR TITLE
grpc_json_transcoder: expose underlying JsonPrintOptions

### PR DESCRIPTION
**Pull request by jwfang**
1. pump to https://github.com/grpc-ecosystem/grpc-httpjson-transcoding/commit/e4f58aa07b9002befa493a0a82e10f2e98b51fc6 for *JsonPrintOptions*
2. expose *JsonPrintOptions* as *print_options* config;
3. docs

cc @lizan

### CODE REVIEW

**Comment by lizan** [comment]
Thanks for adding the JSON Schema and plumbing this.
Can you add some test cases for each field of JsonPrintOptions?

**Comment by mattklein123** [comment]
@lizan please tag the rest of us when you are done with first round of review. Thanks.

File: test/common/grpc/json_transcoder_filter_test.cc [hunk 1]
```diff
@@ -285,5 +285,90 @@ TEST_F(GrpcJsonTranscoderFilterTest, TranscodingUnaryError) {
...
+  EXPECT_EQ(GetParam().expected_response_, response_json);
+}
+
+INSTANTIATE_TEST_CASE_P(
+    GrpcJsonTranscoderFilterPrintOptions, GrpcJsonTranscoderFilterPrintTest,
+    ::testing::Values(
+        GrpcJsonTranscoderFilterPrintTestParam{
+            R"(
+    {
+     "proto_descriptor": "test/proto/bookstore.descriptor",
```
Comment (lizan) [inline test/common/grpc/json_transcoder_filter_test.cc]: will the path work without `TestEnvironment::runfilesPath`?

File: test/common/grpc/json_transcoder_filter_test.cc [hunk 2]
```diff
@@ -285,5 +285,90 @@ TEST_F(GrpcJsonTranscoderFilterTest, TranscodingUnaryError) {
...
+    }
+			                                           )",
+            R"({"id":"101","gender":"MALE","lname":"Shakespeare"})"},
+        GrpcJsonTranscoderFilterPrintTestParam{R"(
+    {
+     "proto_descriptor": "test/proto/bookstore.descriptor",
+     "services": ["bookstore.Bookstore"],
+     "print_options":{"add_whitespace": true}
+    }
+			                                           )",
```
Comment (lizan) [inline test/common/grpc/json_transcoder_filter_test.cc]: nit: the style here looks weird, maybe you can have `{` on the same line as `R"(`, and close on same line too.

**Comment by lizan** [comment]
Almost LGTM, @mattklein123 @htuch for additional review.

File: docs/configuration/http_filters/grpc_json_transcoder_filter.rst [hunk 3]
```diff
@@ -51,6 +57,30 @@ proto_descriptors
...
+  `JsonPrintOptions <https://developers.google.com/protocol-buffers/docs/reference/cpp/
+  google.protobuf.util.json_util#JsonPrintOptions>`. Valid options are:
+
+  add_whitespace
+    *(optional, boolean)* Whether to add spaces, line breaks and indentation to make the JSON
+    output easy to read. Default to false.
+
+  always_print_primitive_fields
+    *(optional, boolean)* Whether to always print primitive fields. By default primitive fields
+    with default values will be omitted in JSON joutput. For example, an int32 field set to 0
```
Comment (htuch) [inline docs/configuration/http_filters/grpc_json_transcoder_filter.rst]: Nit: s/joutput/output/

File: source/common/grpc/json_transcoder_filter.cc [hunk 4]
```diff
@@ -112,6 +112,15 @@ JsonTranscoderConfig::JsonTranscoderConfig(const Json::Object& config) {
   type_helper_.reset(new google::grpc::transcoding::TypeHelper(
       google::protobuf::util::NewTypeResolverForDescriptorPool(TYPE_URL_PREFIX,
                                                                &descriptor_pool_)));
+
+  auto print_config = config.getObject("print_options", true);
```
Comment (htuch) [inline source/common/grpc/json_transcoder_filter.cc]: Nit: `const auto` (if the compiler will allow it).

File: test/common/grpc/json_transcoder_filter_test.cc (hunk 2 above)
Comment (htuch) [inline test/common/grpc/json_transcoder_filter_test.cc]: +1

File: test/common/grpc/json_transcoder_filter_test.cc (hunk 1 above)
Comment (htuch) [inline test/common/grpc/json_transcoder_filter_test.cc]: There is also the `{{ test_rundir }}` substitution available at https://github.com/lyft/envoy/blob/master/test/test_common/environment.cc#L112

File: test/common/grpc/json_transcoder_filter_test.cc [hunk 5]
```diff
@@ -322,52 +324,41 @@ INSTANTIATE_TEST_CASE_P(
...
      "print_options":{"add_whitespace": true}
-    }
-			                                           )",
+    })",
                                                R"({
  "id": "101",
  "gender": "MALE",
  "lname": "Shakespeare"
 }
 )"},
```
Comment (jwfang) [inline test/common/grpc/json_transcoder_filter_test.cc]: this extra new-line can't be collpase since *add_spaces* will output a trailing new-line.

**Comment by jwfang** [comment]
@lizan @htuch thanks for your review and help. you are really quick. :)

i found the following error message from *check_format.py* a bit misleading when the file is indeed depend on //source/common/protobuf but still using *google::protobuf* namespace:

> ERROR: ./source/common/grpc/json_transcoder_filter.h has unexpected direct external dependency on protobuf, use //source/common/protobuf instead.

it's no big deal and may rare, but it would be more helpful if it says "replace google::protobuf with Protobuf".

**Pull request merged**